### PR TITLE
refactored gcsFs and ghFs tests

### PR DESCRIPTION
**Pull request by singhravidutt**
Refactored code which was added to bypass the shortcoming of https://github.com/junit-team/junit4/issues/1509

### CODE REVIEW

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 1]
```diff
@@ -100,34 +99,38 @@ public class GoogleHadoopFileSystemIntegrationTest extends GoogleHadoopFileSyste
...
+  }
 
-        @Override
-        public void after() {
-          GoogleHadoopFileSystemTestBase.storageResource.after();
-        }
-      };
+  @After
+  public void after() throws IOException {
+    if (gcs != null) {
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:125]: Should it check `gcsiHelper` for `null` here?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 2]
```diff
@@ -1054,7 +1057,7 @@ public void testGlobStatus() throws IOException {
     createFile(new Path("/directory1/subdirectory2/file1"), data);
     createFile(new Path("/directory1/subdirectory2/file2"), data);
 
-    FileStatus[] rootDirectories = ghfs.globStatus(new Path("/d*"));
+    FileStatus[] rootDirectories = ghfs.globStatus(new Path("/directory1*"));
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:1060]: Why do we need to change glob pattern?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java [hunk 3]
```diff
@@ -33,38 +33,40 @@
...
-import org.junit.AfterClass;
-import org.junit.BeforeClass;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
 @RunWith(JUnit4.class)
-public final class GoogleHadoopFileSystemXAttrsIntegrationTest {
+public final class GoogleHadoopFileSystemXAttrsIntegrationTest extends HadoopFileSystemTestBase {
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java:44]: This seems out of scope of this PR? Also, I do not think that we need to follow inheritance pattern in tests that do not use it.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java [hunk 4]
```diff
@@ -275,6 +277,21 @@ public void removeXAttr() throws Exception {
...
+          /**
+           * Returns the MethodOutcome of trying to rename an existing file into the root directory.
+           */
+          @Override
+          public MethodOutcome renameFileIntoRootOutcome() {
+            return new MethodOutcome(MethodOutcome.Type.RETURNS_TRUE);
+          }
+        });
+  }
+
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java:294]: This seems out of scope for PR that reverts hack after JUnit fix.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemIntegrationTest.java [hunk 5]
```diff
@@ -60,44 +60,34 @@ public class HadoopFileSystemIntegrationTest extends HadoopFileSystemTestBase {
...
+    postCreateInit();
+    ghfsHelper.setIgnoreStatistics(); // Multi-threaded code screws us up.
+  }
 
   /** Perform initialization after creating test instances. */
-  public static void postCreateInit() throws IOException {
-    HadoopFileSystemTestBase.postCreateInit();
+  @Override
+  public void postCreateInit() throws IOException {
+    super.postCreateInit();
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemIntegrationTest.java:90]: Should inheritance take care of this automatically?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationHelper.java [hunk 6]
```diff
@@ -45,6 +46,10 @@ public static GoogleCloudStorageFileSystem createGcsFs(String projectId) throws
             .build());
   }
 
+  public static ClientType getClientType(boolean testStorageClient) {
+    return testStorageClient ? ClientType.STORAGE_CLIENT : ClientType.HTTP_API_CLIENT;
+  }
+
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationHelper.java:52]: Seem to be out of scope of this PR.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageNewIntegrationTest.java [hunk 7]
```diff
@@ -692,7 +692,9 @@ public void listObjectInfo_includePrefix_objectInBucket() throws Exception {
   public void listObjectInfo_includePrefix_implicitDirInBucket() throws Exception {
     GoogleCloudStorage gcs = createGoogleCloudStorage(gcsOptions);
 
-    String testBucket = gcsfsIHelper.createUniqueBucket("lst-objs_incl-pfx_impl-dir-in-bckt");
+    String testBucket =
+        gcsfsIHelper.createUniqueBucket(
+            UUID.randomUUID().toString().replaceAll("-", "").substring(0, 10));
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageNewIntegrationTest.java:697]: Out of scope of PR that reverts hack after JUnit fix.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageIntegrationHelper.java [hunk 8]
```diff
@@ -49,7 +49,7 @@
 public abstract class GoogleCloudStorageIntegrationHelper {
 
   // Prefix used for naming test buckets.
-  private static final String TEST_BUCKET_NAME_PREFIX = "gcsio-test";
+  private static final String TEST_BUCKET_NAME_PREFIX = "dataproc-gcs-gcsio";
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageIntegrationHelper.java:52]: Out of scope

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java [hunk 9]
```diff
@@ -210,7 +177,7 @@ private void validateFileInfoInternal(
       assertWithMessage(
               "getModificationTime for bucketName '%s' objectName '%s'", bucketName, objectName)
           .that(fileModificationTime)
-          .isAtLeast(testStartTime);
+          .isAtLeast(Instant.ofEpochMilli(testStartTime.toEpochMilli()));
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java:180]: Why do we need this change?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java [hunk 10]
```diff
@@ -51,117 +51,84 @@
...
 
   // Time when test started. Used for determining which objects got
   // created after the test started.
-  protected static Instant testStartTime;
+  protected Instant testStartTime;
 
-  protected static String sharedBucketName1;
-  protected static String sharedBucketName2;
+  public String sharedBucketName1;
+  public String sharedBucketName2;
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java:79]: Do we need these to be public?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 11]
```diff
@@ -1474,7 +1477,7 @@ private void testGlobStatusFlatConcurrent(GlobAlgorithm globAlgorithm) throws IO
     ghfs.mkdirs(new Path("/directory1/subdirectory1"));
     createFile(new Path("/directory1/subdirectory1/file1"), "data".getBytes(UTF_8));
 
-    FileStatus[] rootDirStatuses = ghfs.globStatus(new Path("/d*"));
+    FileStatus[] rootDirStatuses = ghfs.globStatus(new Path("/directory1*"));
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:1480]: Why do we need to change pattern?

**Review by medb — CHANGES_REQUESTED** [review]

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 2 above)
Comment (singhravidutt) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:1060]: It's because we updated the bucket prefix. As it's going to be removed from this PR this change will also go away

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemIntegrationTest.java (hunk 5 above)
Comment (singhravidutt) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemIntegrationTest.java:90]: yeah, will remove it.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationHelper.java (hunk 6 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationHelper.java:52]: yeah. Will do the cleanup.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageNewIntegrationTest.java (hunk 7 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageNewIntegrationTest.java:697]: yeah. Will do the cleanup.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageIntegrationHelper.java (hunk 8 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageIntegrationHelper.java:52]: yeah. Will do the cleanup.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 11 above)
Comment (singhravidutt) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:1480]: Out of scope. yeah. Will do the cleanup.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java (hunk 10 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java:79]: protected will do. Will update.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java (hunk 9 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java:180]: It was flaky and caused issues. Reason behind why it was flaky was that, for InMemory gcs operation completes in nano seconds. We might not require this now. Will check and update

Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java:180]: So, we need this change specifically for Unit tests where gcs operation will just complete in nano second. This wasn't an issue earlier because `testStartTime` used to get initialized only once in `@BeforeClass` now it will be `initialized` for every test.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java [hunk 12]
```diff
@@ -33,38 +35,71 @@
...
+    ghfs.initialize(initUri, GoogleHadoopFileSystemTestBase.loadConfig());
+
+    gcsiHelper =
+        new GoogleCloudStorageFileSystemIntegrationHelper(
+            ((GoogleHadoopFileSystem) ghfs).getGcsFs());
+    ghfs.mkdirs(new Path(ghfs.getUri()));
+
+    ghfsHelper = new HadoopFileSystemIntegrationHelper(ghfs);
+    gcsiHelper.beforeAllTests();
+    bucketName = gcsiHelper.sharedBucketName1;
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java:77]: It seems like this snippet of code is used in many places, should we extract it in the helper?

**Review by medb — APPROVED** [review]

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java (hunk 9 above)
Comment (medb) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java]: Should we initialize `testStartTime` in `@BeforeClass` instead, then?

Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java]: Updated

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java (hunk 12 above)
Comment (singhravidutt) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemXAttrsIntegrationTest.java:77]: Will take care of it during refactoring integration tests further.
